### ISSUE TITLE
502 Bad Gateway when accepting Developer Programme Agreement

### ISSUE THREAD

**Issue by tylerl0706** [canonical-websites/snapcraft.io#1040]
### Expected behaviour

I go to the dashboard.

### Steps to reproduce the problem

Go to https://snapcraft.io
Go to "My Published Snaps" in top right drop down menu

Note: it redirects to https://snapcraft.io/account/agreement

click "I agree" and submit.

Get:
<img width="354" alt="image" src="https://user-images.githubusercontent.com/2644648/44621463-b2cbef80-a85b-11e8-8bab-4649c626b65e.png">

### Specs

- _URL:_ https://snapcraft.io/account/agreement

- _Operating system:_ macOS Mojave
- _Browser:_ Safari 12.0 & Chrome

**Comment by tbille** [comment]
@tylerl0706 
Thank you for reporting this issue.
It seems that I can't reproduce this error, could you confirm me you still have this error?

**Comment by tylerl0706** [comment]
I can still repro this. Let me know how I can help.

This initially started with a coworker of mine adding me to a snap package's owners (or maintainers... Not sure what they're officially called). I don't think I had an account, prior.

**Comment by tylerl0706** [comment]
I'm still running into this. I can't be added as an owner of the package until this is addressed.

**Comment by tylerl0706** [comment]
Just tried again on my phone and now "continue" is grayed out:

![screenshot_20180904-102942](https://user-images.githubusercontent.com/2644648/45047321-92efb500-b02d-11e8-9ab7-c38309a0dc7f.jpg)

**Comment by tbille** [comment]
It seems that we have a strange behaviour on your account, we need to investigate on it.
In the meantime you can use our old webview to manage your snaps: https://dashboard.snapcraft.io/

**Comment by tylerl0706** [comment]
Unfortunately, I get a strange behavior there as well when I try to go to my account details:

![screenshot_20180905-115016__01](https://user-images.githubusercontent.com/2644648/45114620-517d0980-b102-11e8-9e47-8f5a6a4d93d6.jpg)
![screenshot_20180905-115047](https://user-images.githubusercontent.com/2644648/45114634-5a6ddb00-b102-11e8-9252-e57ac7041a8e.jpg)

**Comment by 3jnsn** [comment]
I had the second issue, where 'continue' stayed inactive even after clicking to accept the terms.

@tbille 's suggest to go to https://dashboard.snapcraft.io/ instead seemed to solve it, but..

following this tutorial: https://docs.snapcraft.io/build-snaps/pre-built

everything seemed to go fine until I tried to reserve a package name. the following happens:

> host:~$ sudo snap login <email>
> Personal information is handled as per our privacy notice at
> https://www.ubuntu.com/legal/dataprivacy/snap-store
> 
> Password of "<email>": 
> Login successful
> host:~$ sudo snapcraft register geekbench-test-user
> 
> We always want to ensure that users get the software they expect
> for a particular name.
> .
> .
> you, and be the software you intend to publish there? [y/N]: y
> Registering geekbench-test-user.
> No valid credentials found. Have you run "snapcraft login"?
> Invalid credentials: Root macaroon not in the config file. Have you run "snapcraft login"?

after successful login, it behaves as if I'm not logged in. I've tried logging out/back in many times, same result. 

could it be related to the original auth issue? 

I'm planning to create a new issue, but thought I'd start here since it seems possible the two issues are related. 

a few more thoughts

1. when I first registered I was on macOS (HS)/Chrome
2. the email verification function doesn't seem to be able to handle .co domains. I created the account with a .co email, then had add my gmail for verification to work. no idea if it's related. 
3. I was running snapcraft on Ubuntu 18 in VirtualBox

**Comment by sparkiegeek** [comment]
After logging in with `sudo snapcraft login` you no longer need to use `sudo` with snapcraft. Can you try a plain `snapcraft register` and see if you still have the issue?

**Comment by tylerl0706** [comment]
I tried doing a `snapcraft login` to see if that would get me past the issue and it sadly didn't:

```
/home/tylerleonhardt
~> snapcraft login
Enter your Ubuntu One e-mail address and password.
If you do not have an Ubuntu One account, you can create one at https://dashboard.snapcraft.io/openid/login
Email: tylerl0706@gmail.com
Password: 

We strongly recommend enabling multi-factor authentication: https://help.ubuntu.com/community/SSO/FAQs/2FA
Do you agree to the developer terms and conditions. (https://dashboard.snapcraft.io/dev/agreements/new/)? [y/N] y
Unexpected error encountered during signing the developer terms and conditions. Please visit https://dashboard.snapcraft.io/dev/agreements/new/ and agree to the terms and conditions before continuing.
Login failed.
```
I feel like my account got into a bad state...

**Comment by 3jnsn** [comment]
@sparkiegeek thanks for the suggestion. unfortunately the result is the same with or without sudo

I did get a step further by registering my snap here instead: https://dashboard.snapcraft.io/register-snap/

but ran into the same issue again when I tried to push it from the command line. I've submitted this for my snapcraft issue:

https://bugs.launchpad.net/snapcraft/+bug/1791879

@tylerl0706 seems like we have different issues. I didn't intend to threadjack, thought they might be related. 

one parting thought, is there anything odd about your account in terms of the email addresses? do you have more than one? does your email use a domain other than .com? my problems first started when I never received a confirmation email to my main address which is .co. 

good luck!

**Comment by janoskk** [comment]
@3jnsn I think it's not related to the .co email address. I have the same issue and I also can't solve it.

I believe it's browser independent, I tried on Win/Mac/Linux with the most recent Chrome and FF. The gray Continue button can be activated from the inspector, but Bad Gateway or Too many redirects responses cannot be avoided.

These pages response the ERR_TOO_MANY_REDIRECTS:
https://dashboard.snapcraft.io/dev/account/
https://dashboard.snapcraft.io/dev/agreements/new/

When trying to login and accept with `snapcraft login` I got the same error as @tylerl0706. Snapcraft version is 2.43 (stable).

**Comment by tylerl0706** [comment]
@tbille That did the trick!! Thank you so much.

**Comment by tbille** [comment]
@tylerl0706 no problem, this was a tricky one :smile:

**Comment by janoskk** [comment]
@tbille Sorry but I already set the username. Could you check our account? It's central-devops at logmein.com and the username is logmein.

**Comment by cprov** [comment]
@janoskk I see that you unblocked yourself today, registered and upload new revisions of https://snapcraft.io/logmein-host. Let us know if we missed anything on our previous instructions.

**Comment by janoskk** [comment]
Finally, I was able to "solve" my issue. Here is what I did, each step in a new incognito window to avoid mixing the accounts.

1. I removed my account. Before that, renamed to something else to avoid any collision in the future.
2. Registered a new account initiated at login.ubuntu.com, *not* at snapcraft.io. Note that reusing the removed account's primary email address was not OK and led to 502 Bad Gateway again.
3. Logged in at dashboard.snapcraft.io (*not* anywhere else) and pressed the "Add your snap" button. Then I was able to accept the Developer Programme Agreement.
4. Added the removed account's primary email to the new account's emails and made it primary at login.ubuntu.com.
